src/runtime/webgpu.cpp

```diff
@@ -0,0 +1,1146 @@
+#include "HalideRuntimeWebGPU.h"
+#include "device_buffer_utils.h"
+#include "device_interface.h"
+#include "gpu_context_common.h"
+#include "printer.h"
+#include "scoped_spin_lock.h"
+
+#include "mini_webgpu.h"
+
+namespace Halide {
+namespace Runtime {
+namespace Internal {
+namespace WebGPU {
+
+extern WEAK halide_device_interface_t webgpu_device_interface;
+
+WEAK int create_webgpu_context(void *user_context);
+
+// A WebGPU instance/adapter/device defined in this module with weak linkage.
+WEAK WGPUInstance global_instance = nullptr;
+WEAK WGPUAdapter global_adapter = nullptr;
+WEAK WGPUDevice global_device = nullptr;
+// Lock to synchronize access to the global WebGPU context.
+volatile ScopedSpinLock::AtomicFlag WEAK context_lock = 0;
+
+// Size of the staging buffer used for host<->device copies.
+constexpr int kWebGpuStagingBufferSize = 4 * 1024 * 1024;
+// A staging buffer used for host<->device copies.
+WGPUBuffer WEAK staging_buffer = nullptr;
+
+}  // namespace WebGPU
+}  // namespace Internal
+}  // namespace Runtime
+}  // namespace Halide
+
+using namespace Halide::Runtime::Internal;
+using namespace Halide::Runtime::Internal::WebGPU;
+
+extern "C" {
+// TODO: Remove all of this when wgpuInstanceProcessEvents() is supported.
+// See https://github.com/halide/Halide/issues/7248
+#ifdef WITH_DAWN_NATIVE
+// Defined by Dawn, and used to yield execution to asynchronous commands.
+void wgpuDeviceTick(WGPUDevice);
+// From <unistd.h>, used to spin-lock while waiting for device initialization.
+int usleep(uint32_t);
+#else
+// Defined by Emscripten, and used to yield execution to asynchronous Javascript
+// work in combination with Emscripten's "Asyncify" mechanism.
+void emscripten_sleep(unsigned int ms);
+// Wrap emscripten_sleep in wgpuDeviceTick() to unify usage with Dawn.
+void wgpuDeviceTick(WGPUDevice) {
+    emscripten_sleep(1);
+}
+#endif
+
+// The default implementation of halide_webgpu_acquire_context uses the global
+// pointers above, and serializes access with a spin lock.
+// Overriding implementations of acquire/release must implement the following
+// behavior:
+// - halide_webgpu_acquire_context should always store a valid
+//   instance/adapter/device in instance_ret/adapter_ret/device_ret, or return
+//   an error code.
+// - A call to halide_webgpu_acquire_context is followed by a matching call to
+//   halide_webgpu_release_context. halide_webgpu_acquire_context should block
+//   while a previous call (if any) has not yet been released via
+//   halide_webgpu_release_context.
+WEAK int halide_webgpu_acquire_context(void *user_context,
+                                       WGPUInstance *instance_ret,
+                                       WGPUAdapter *adapter_ret,
+                                       WGPUDevice *device_ret,
+                                       bool create = true) {
+    halide_abort_if_false(user_context, &context_lock != nullptr);
```

Comment:
There's lots of calls to `halide_abort_if_false()` in existing code, but IMHO we should really avoid it in new code (and ~most of the existing usage should probably be changed) -- we really don't want to `abort()` in runtime code unless we know the error is **unrecoverable**.  In this specific case, someone would have to override the WEAK `context_lock` symbol to make it a null value, which is in the category of "don't do that" -- i.e., I don't see why we should ever check for that sort of pathological condition at runtime.

Reply:
> Should I replace them all with halide_debug_assert()

I'll leave that up to you -- what you have is no worse than other backends currently, it's just that we have developed an antipattern. (I have on my task list "go examine all the halide_abort_if_false and convert to either halide_debug_assert or actual error", but that's gonna be a gradual background task most likely.)

>  should we not abort even in debug mode?

I'd argue that for the Halide runtime, we shouldn't ever call abort() unless we know the system is in a compromised state (e.g. memory corruption). (Note that the default implementation of `halide_error()` *does* call abort(), but it's easy to override and almost always done so for production software.) So, it's a judgement call on your part, but for this backend, I'm guessing there aren't a lot of obvious places where abort() is the right answer.